File: src/main/java/raccoonlang/parser/ast/FunctionParameterAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.Token;
import raccoonlang.tokenizer.TokenType;

public class FunctionParameterAstNode {
    public FqtnAstNode typeName;
    public Token parameterName;

    public static FunctionParameterAstNode parse(Parser parser) {
        FunctionParameterAstNode functionParameterAstNode = new FunctionParameterAstNode();
        functionParameterAstNode.typeName = FqtnAstNode.parse(parser);
        functionParameterAstNode.parameterName = parser.take(TokenType.IDENTIFIER);
        return functionParameterAstNode;
    }

    @Override
    public String toString() {
        return "FunctionParameterAstNode{" +
                "typeName=" + typeName.toString() +
                ", parameterName=" + parameterName.toString() +
                '}';
    }
}

File: Parser/AST/FunctionParameterAstNode.cs
namespace Raccoonlang.Parser.AST;

using Tokenizer;

public class FunctionParameterAstNode
{
    public FqtnAstNode? paramType;
    public Token? paramName;

    public static FunctionParameterAstNode Parse(Parser parser)
    {
        FunctionParameterAstNode node = new FunctionParameterAstNode();
        node.paramType = FqtnAstNode.Parse(parser);
        node.paramName = parser.Take(TokenType.IDENTIFIER);
        return node;
    }

    public override string ToString()
    {   
        if (paramType != null && paramName != null) {
            return "FunctionParameterAstNode{" +
                    "paramType=" + paramType.ToString() +
                    ", paramName=" + paramName.ToString() +
                    "}";
        } else {
            return "FunctionParameterAstNode{}";
        }
    }
}

public class FunctionParametersAstNode
{
    public List<FunctionParameterAstNode> paramList = new List<FunctionParameterAstNode>();

    public static FunctionParametersAstNode Parse(Parser parser)
    {
        FunctionParametersAstNode node = new FunctionParametersAstNode();

        parser.Take(TokenType.OPEN_PAREN);

        if (parser.Peek().Type != TokenType.CLOSE_PAREN) {
            node.paramList.Add(FunctionParameterAstNode.Parse(parser));
            
            while(parser.Peek().Type == TokenType.COMMA) {
                parser.Take();
                node.paramList.Add(FunctionParameterAstNode.Parse(parser));
            }
        }

        parser.Take(TokenType.CLOSE_PAREN);
        return node;
    }

    public override string ToString()
    {
        return "FunctionParametersAstNode{" +
                "paramList=" + paramList.ToArray().ToString() +
                "}";
    }
}